Bugzilla/App/Plugin/BlockIP.pm

```diff
@@ -38,18 +67,79 @@ sub _unblock_ip {
   $MEMCACHED->delete("block_ip:$ip") if $MEMCACHED;
 }
 
+sub _check_ip {
+  my ($c, $ip) = @_;
+
+  state $iprepd_url = Bugzilla->localconfig->iprepd_url;
+  state $iprepd_key = Bugzilla->localconfig->iprepd_key;
+
+  if ($MEMCACHED && $MEMCACHED->get("block_ip:$ip")) {
+    return 'block_internal';
+  }
+
+  if ($iprepd_url && $iprepd_key) {
+    my $params = Bugzilla->params;
+    my $path   = Mojo::Path->new('/type/ip/')->merge($ip);
+    my $url    = Mojo::URL->new($iprepd_url)->path($path);
+
+    return '' unless $params->{iprepd_active};
+
+    my $filter = Bugzilla::Bloomfilter->lookup("rate_limit_whitelist");
+    if ($filter->test($ip)) {
+      INFO("$ip is in the rate limit whitelist, ignoring iprepd.");
+      return '';
+    }
+
+    my $tx   = $c->ua->get($url => {Authorization => "APIKey $iprepd_key"});
```

Comment:
I'm not sure if it is possible with the library being used, but one feature that might be useful is a configuration option specifying a maximum end-to-end request timeout for iprepd queries; if the request exceeds this it would just pass. This would prevent any sort of service impact associated with excessive latency.

Reply:
The mojo framework has defaults for this that are reasonable. Timeouts will behave as pseudo-http responses and be logged by as errors.

Reply:
(There are two separate limits, connection timeout and inactivity timeout)

Additionally the API is non-blocking so only the current request is impacted and the total number of concurrent requests is unaffected)